## Akshit Ram Pershad
### U87835413

# Unit 1 Assignment

In this assignment, we will focus on sports analytics. This data set is made available by http://www.baseball-reference.com. It contains data about professional baseball (MLB) games played in the 2016 season. There are 2,427 games in the data set. Each row represents a single game. The goal is to predict the attendance at a home team’s game. This is an important task because most franchises want to predict the number of attendees for a variety of reasons including profits.

## Description of Variables

The description of variables are provided in "Baseball - Data Dictionary.docx"

## Goal

Use the **baseball.csv** data set and build a model to predict **attendance_binary**.

## Submission:

Please save and submit this Jupyter notebook file. The correctness of the code matters for your grade. **Readability and organization of your code is also important.** You may lose points for submitting unreadable/undecipherable code. Therefore, use markdown cells to create sections, and use comments where necessary.


# Section 1: (6 points in total)

## Data Prep (5.5 points)

# Setup

In [1703]:
# Common imports
import numpy as np
import pandas as pd

np.random.seed(42)

# Get the data

In [1704]:
baseball = pd.read_csv("baseball.csv")
baseball.head()

attendance_binary  previous_attendance  previous_away_team_errors  \
0                  0                43683                          2   
1                  0                45785                          0   
2                  0                48282                          0   
3                  0                21830                          0   
4                  0                49289                          2   

   previous_away_team_hits  previous_away_team_runs   game_type  \
0                        6                        2  Night Game   
1                        7                        2  Night Game   
2                        8                        4  Night Game   
3                        9                        6    Day Game   
4                        4                        2  Night Game   

  previous_game_type  previous_home_team_errors  previous_home_team_hits  \
0           Day Game                          0                        6   
1           Day Game                          0                       10   
2           Day Game                          2                        4   
3         Night Game                          0                       15   
4           Day Game                          1                        1   

   previous_home_team_runs   game_day previous_game_day  temperature  \
0                        6  Wednesday            Monday           55   
1                        3  Wednesday            Monday           48   
2                        3  Wednesday            Monday           65   
3                       11  Wednesday           Tuesday           77   
4                        3    Tuesday            Monday           81   

   wind_speed       sky  previous_game_duration  previous_homewin  
0          24  Overcast                2.933333                 1  
1           7   Unknown                2.800000                 1  
2          10    Cloudy                3.383333                 0  
3           0   In Dome                3.233333                 1  
4          12    Cloudy                2.633333                 1

# Split the data into train and test 70/30

In [1705]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(baseball, test_size=0.3)

## Check the missing values in Train and Test

In [1706]:
train.isna().sum()

attendance_binary            0
previous_attendance          0
previous_away_team_errors    0
previous_away_team_hits      0
previous_away_team_runs      0
game_type                    0
previous_game_type           0
previous_home_team_errors    0
previous_home_team_hits      0
previous_home_team_runs      0
game_day                     0
previous_game_day            0
temperature                  0
wind_speed                   0
sky                          0
previous_game_duration       0
previous_homewin             0
dtype: int64

In [1707]:
test.isna().sum()

attendance_binary            0
previous_attendance          0
previous_away_team_errors    0
previous_away_team_hits      0
previous_away_team_runs      0
game_type                    0
previous_game_type           0
previous_home_team_errors    0
previous_home_team_hits      0
previous_home_team_runs      0
game_day                     0
previous_game_day            0
temperature                  0
wind_speed                   0
sky                          0
previous_game_duration       0
previous_homewin             0
dtype: int64

# Data Prep

In [1708]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

# Separating the target variable and input variables

In [1709]:
train_targets = train[[ 'attendance_binary']] 
test_targets = test[[ 'attendance_binary']]

train_inputs = train.drop(['attendance_binary'],  axis=1)
test_inputs = test.drop(['attendance_binary'] , axis=1)

In [1710]:
train_inputs

previous_attendance  previous_away_team_errors  previous_away_team_hits  \
270                 19072                          1                        8   
2096                40725                          0                        6   
1055                34036                          1                        4   
1014                27134                          1                       14   
1320                41571                          0                        8   
...                   ...                        ...                      ...   
1638                32946                          1                        8   
1095                40419                          0                        9   
1130                25011                          1                        6   
1294                39594                          0                       11   
860                 20458                          2                        7   

      previous_away_team_runs   game_type previous_game_type  \
270                         6  Night Game         Night Game   
2096                        4    Day Game         Night Game   
1055                        1  Night Game         Night Game   
1014                        7  Night Game         Night Game   
1320                        2  Night Game           Day Game   
...                       ...         ...                ...   
1638                        1  Night Game           Day Game   
1095                        2    Day Game         Night Game   
1130                        0    Day Game         Night Game   
1294                        8    Day Game         Night Game   
860                         2  Night Game         Night Game   

      previous_home_team_errors  previous_home_team_hits  \
270                           1                       11   
2096                          0                       10   
1055                          0                        6   
1014                          0                       10   
1320                          1                        8   
...                         ...                      ...   
1638                          0                        4   
1095                          1                        9   
1130                          1                       12   
1294                          0                        7   
860                           0                        4   

      previous_home_team_runs   game_day previous_game_day  temperature  \
270                         7     Friday          Thursday           76   
2096                        5   Thursday         Wednesday           61   
1055                        5   Saturday            Friday           84   
1014                        5  Wednesday           Tuesday           91   
1320                        4     Sunday          Saturday           68   
...                       ...        ...               ...          ...   
1638                        2    Tuesday            Sunday           83   
1095                        3  Wednesday           Tuesday           88   
1130                       11     Sunday          Saturday           78   
1294                        4     Sunday          Saturday           77   
860                         4    Tuesday            Monday           80   

      wind_speed      sky  previous_game_duration  previous_homewin  
270            7   Cloudy                3.100000                 1  
2096          16   Cloudy                3.050000                 1  
1055          10   Cloudy                2.433333                 1  
1014          11   Cloudy                3.083333                 0  
1320          17    Sunny                3.050000                 1  
...          ...      ...                     ...               ...  
1638           6   Cloudy                2.816667                 1  
1095          15    Sunny                2.800000                 1  
1130        

In [1711]:
train_x.shape

(1698, 19)

In [1712]:
train_targets

attendance_binary
270                   0
2096                  1
1055                  1
1014                  0
1320                  1
...                 ...
1638                  1
1095                  1
1130                  0
1294                  1
860                   0

[1698 rows x 1 columns]

In [1713]:
train_targets.shape

(1698, 1)

##  Identify the numerical, binary and categorical columns manually 

In [1714]:
numeric_columns = ['previous_attendance','previous_away_team_errors', 'previous_away_team_hits','previous_away_team_runs',
                   'previous_home_team_errors', 'previous_home_team_hits', 'previous_home_team_runs', 'temperature',
                   'wind_speed','previous_game_duration']
 
binary_columns = ['previous_homewin']
 
categorical_columns = ['sky']

In [1715]:
numeric_columns

['previous_attendance',
 'previous_away_team_errors',
 'previous_away_team_hits',
 'previous_away_team_runs',
 'previous_home_team_errors',
 'previous_home_team_hits',
 'previous_home_team_runs',
 'temperature',
 'wind_speed',
 'previous_game_duration']

In [1716]:
binary_columns

['previous_homewin']

In [1717]:
categorical_columns

['sky']

# Pipeline

In [1718]:
# Numeric transformer:

numeric_transformer = Pipeline(steps=[
                ('imputer', SimpleImputer(strategy='median')),
                ('scaler', StandardScaler())])

In [1719]:
# Categorical transformer:

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='unknown')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])


In [1720]:
# Binary transformer:

binary_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent'))])



In [1721]:
# Column transformer:

preprocessor = ColumnTransformer([
        ('num', numeric_transformer, numeric_columns),
        ('cat', categorical_transformer, categorical_columns),
        ('binary', binary_transformer, binary_columns)],
        remainder='drop')

## Performing Fit and transform for the train data

In [1722]:
train_x = preprocessor.fit_transform(train_inputs)
train_x


array([[-1.12371621,  0.57666325, -0.20719118, ...,  0.        ,
         0.        ,  1.        ],
       [ 1.05787587, -0.72716391, -0.78017264, ...,  0.        ,
         0.        ,  1.        ],
       [ 0.38394295,  0.57666325, -1.35315411, ...,  0.        ,
         0.        ,  1.        ],
       ...,
       [-0.52534761,  0.57666325, -0.78017264, ...,  0.        ,
         0.        ,  1.        ],
       [ 0.94392488, -0.72716391,  0.65228102, ...,  0.        ,
         0.        ,  0.        ],
       [-0.98407336,  1.88049041, -0.49368191, ...,  1.        ,
         0.        ,  1.        ]])

In [1723]:
train_x.shape

(1698, 19)

## Performing Transform for the test data

In [1724]:
test_x = preprocessor.transform(test_inputs)
test_x

array([[ 0.0814842 ,  1.88049041, -1.06666338, ...,  1.        ,
         0.        ,  1.        ],
       [ 1.17001331, -0.72716391, -1.63964484, ...,  0.        ,
         0.        ,  0.        ],
       [-0.29905768, -0.72716391, -0.20719118, ...,  0.        ,
         1.        ,  0.        ],
       ...,
       [-0.49582715, -0.72716391,  1.22526249, ...,  0.        ,
         0.        ,  0.        ],
       [ 0.26445059, -0.72716391, -0.20719118, ...,  0.        ,
         0.        ,  1.        ],
       [-1.71775245, -0.72716391, -1.63964484, ...,  0.        ,
         0.        ,  0.        ]])

In [1725]:
test_x.shape

(729, 19)

## Find the Baseline (0.5 point)

In [1726]:
from sklearn.dummy import DummyClassifier

dummy_clf = DummyClassifier(strategy="most_frequent")

dummy_clf.fit(train_x, train_targets)

DummyClassifier(strategy='most_frequent')

In [1727]:
from sklearn.metrics import accuracy_score

In [1944]:
#Baseline Train Accuracy
dummy_train_pred = dummy_clf.predict(train_x)

baseline_train_acc = accuracy_score(train_targets, dummy_train_pred)

print('Baseline Train Accuracy: {}' .format(baseline_train_acc))

Baseline Train Accuracy: 0.5141342756183745


In [1945]:
#Baseline Test Accuracy
dummy_test_pred = dummy_clf.predict(test_x)

baseline_test_acc = accuracy_score(test_targets, dummy_test_pred)

print('Baseline Test Accuracy: {}' .format(baseline_test_acc))

Baseline Test Accuracy: 0.5294924554183813


# Section 2: (3 points in total)

Build three different SVM models (by changing the kernels, regularization, etc.). Generate their training and test values. Each model is worth 1 point. 

(Add cells as needed)

## SVM Model 1:

# Train an SVM model with linear kernel


In [1825]:
# Case 2 : Updated default parameters with degree as 3, coef as 2 and C as 10.
           #The Train accuracy is 0.819 and Test accuracy is 0.799
    
from sklearn.svm import SVC
lin_svm = SVC(kernel="linear", degree=3, coef0=3, C=30)
lin_svm.fit(train_x, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=30, coef0=3, kernel='linear')

## Accuracy Check

In [1826]:
from sklearn.metrics import accuracy_score

In [1827]:
#Predict the train values
train_y_pred = lin_svm.predict(train_x)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.8197879858657244

In [1828]:
#Predict the test values
test_y_pred = lin_svm.predict(test_x)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.7997256515775034

#### Therefore, the best best fit for SVM model for linear Kernel is Case 2 when the parameters are set to kernel="linear", degree=3, coef0=3, C=30. which gives the best Accuracy values for train as 0.819 and test as 0.799

## SVM Model 2:

# Train an SVM model with RBF kernel

In [1821]:
## Case 4 : Updated the default parameters with C as 0.1 and gamma as Auto 
        ## The accuracy values of Train is 0.829 and Test is 0.801 
        
rbf_svm = SVC(kernel="rbf", C=0.1, gamma='auto')
rbf_svm.fit(train_x, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=0.1, gamma='auto')

## Accuracy Check

In [1822]:
from sklearn.metrics import accuracy_score

In [1823]:
#Predict the train values
train_y_pred = rbf_svm.predict(train_x)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.8297997644287397

In [1824]:
#Predict the test values
test_y_pred = rbf_svm.predict(test_x)

#Test accuracy
accuracy_score(test_targets, test_y_pred) 

0.8010973936899863

#### Therefore, the best best fit for SVM model for RBK Kernel is Case 4 when the parameters are set to C as 0.1 and gamma as Auto.
#### which gives the best Accuracy values for train as 0.829 and test as 0.801

## SVM Model 3:

# Train an SVM model with poly kernel

In [1856]:
## Case 3 : Updated the parameters with degree as 2, coef as 0.1 and c as 0.1 
        ## The accuracy values of Train is 0.830 and Test is 0.790
    
from sklearn.svm import SVC
pol_svm = SVC(kernel="poly", degree=2, coef0=0.1, C=10)
pol_svm.fit(train_x, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=10, coef0=0.1, degree=2, kernel='poly')

## Accuracy Check

In [1857]:
from sklearn.metrics import accuracy_score

In [1858]:
#Predict the train values
train_y_pred = pol_svm.predict(train_x)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.839811542991755

In [1859]:
#Predict the test values
test_y_pred = pol_svm.predict(test_x)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.7901234567901234

#### Therefore, the best best fit for SVM model for poly Kernel is Case 3, when the parameters are set to kernel="poly", degree=2, coef0=0.1, C=10 which gives the best Accuracy values for train as 0.830 and test as 0.790

#### Hence, Based on the predictions of the all 3 SVM models, RBF gives us the best accuracy value of Test which is 0.80. when the parameters are set to kernel="rbf", C=0.1, gamma='auto'. 

# Section 3: (3 points in total)

Build two different SGD models (by changing the penalty, etc. or adding polynomial terms) and one LogisticRregression model. Generate their training and test values. Each model is worth 1 point.

(Add cells as needed)

## SGD Model 1:
#### In SGD Model 1, we are changing the learning rate to find out the best test accuracy value with penalty as none. 

In [1872]:
## Case 1: max_iter=100, penalty= none, eta0=0.01
## The accuracy values for Train is 0.794 and Test is 0.770

from sklearn.linear_model import SGDClassifier 
        
# tol = stopping criterion
# eta0 = learning rate
# penalty = regularization term
# max_iter = number of passes over training data (i.e., epochs)

sgd_logreg = SGDClassifier(max_iter=100, penalty= None , eta0=0.01)

sgd_logreg.fit(train_x, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SGDClassifier(eta0=0.01, max_iter=100, penalty=None)

# Predicted vs. Actual values

In [1873]:
sgd_logreg.predict(test_x)

array([1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1,

In [1874]:
# Create a new DataFrame

predictions = pd.DataFrame(sgd_logreg.predict(test_x), columns=['Predicted'])
    
predictions

Predicted
0            1
1            1
2            0
3            1
4            1
..         ...
724          1
725          1
726          0
727          1
728          0

[729 rows x 1 columns]

In [1875]:
# Add the actual to the same DataFrame

predictions['Actual'] = np.array(test_targets)
    
predictions

Predicted  Actual
0            1       1
1            1       1
2            0       0
3            1       1
4            1       1
..         ...     ...
724          1       0
725          1       1
726          0       0
727          1       0
728          0       0

[729 rows x 2 columns]

## Accuracy Check

In [1876]:
#Predict the train values
train_y_pred = sgd_logreg.predict(train_x)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.7968197879858657

In [1877]:
#Predict the test values
test_y_pred = sgd_logreg.predict(test_x)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.7709190672153635

#### Therefor, When increasing the learing rate eta0=0.01 their has been and increase in the accuracy value of test and provieded us with best the accuracy value of test i.e 0.770 when compared to the case2 .

# Polynomial Regression with Degree2: 

This is done by creating the polynomial "variables" of the existing variables, then fitting them in a regular regression model


In [1940]:
from sklearn.preprocessing import PolynomialFeatures

# Create second degree terms and interaction terms
poly_features = PolynomialFeatures(degree=2).fit(train_x)

train_x_poly = poly_features.transform(train_x)

test_x_poly = poly_features.transform(test_x)

#Mind you, this will create the polynomial terms of the categorical variables too

#if degree=3, then it creates all combinations: a, a^2, a^3, b, b^2, b^3, a.b, a^2.b, a.b^2, a^2.b^2 

In [1941]:
log_reg = LogisticRegression()

log_reg.fit(train_x_poly, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [1942]:
#Predict the train values
train_y_pred = log_reg.predict(train_x_poly)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.8368669022379269

In [1943]:
#Predict the test values
test_y_pred = log_reg.predict(test_x_poly)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.7832647462277091

## Let's regularize the polynomial regression

In [1888]:
# C is the inverse of alpha. It can only be positive
# Lower C means more regularization

log_reg = LogisticRegression(penalty='l2', 
                              C=0.01)

log_reg.fit(train_x_poly, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=0.01)

In [1889]:
#Predict the train values
train_y_pred = log_reg.predict(train_x_poly)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.8239104829210836

In [1939]:
#Predict the train values
test_y_pred = log_reg.predict(test_x_poly)

#Train accuracy
accuracy_score(test_targets, test_y_pred)

0.7818930041152263

## SGD Model 2:
#### In SGD model 2, we are changing the learning rate and penalty to l1 to find out the best test accuracy value.

In [1897]:
## Case 2 :: max_iter=1000, penalty='l1', eta0=0.03
## The Accuracy values of Train is 0.806 and Test is 0.796

from sklearn.linear_model import SGDClassifier 

# tol = stopping criterion
# eta0 = learning rate
# penalty = regularization term
# max_iter = number of passes over training data (i.e., epochs)

sgd_logreg = SGDClassifier(max_iter=1000, penalty='l1', eta0=0.03) 

sgd_logreg.fit(train_x, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SGDClassifier(eta0=0.03, penalty='l1')

# Predicted vs. Actual values

In [1892]:
sgd_logreg.predict(test_x)

array([1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1,
       1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1,

In [1893]:
# Create a new DataFrame

predictions = pd.DataFrame(sgd_logreg.predict(test_x), columns=['Predicted'])

predictions

Predicted
0            1
1            1
2            0
3            1
4            1
..         ...
724          1
725          1
726          0
727          1
728          0

[729 rows x 1 columns]

In [1894]:
# Add the actual to the same DataFrame

predictions['Actual'] = np.array(test_targets)

predictions

Predicted  Actual
0            1       1
1            1       1
2            0       0
3            1       1
4            1       1
..         ...     ...
724          1       0
725          1       1
726          0       0
727          1       0
728          0       0

[729 rows x 2 columns]

In [1895]:
#Predict the train values
train_y_pred = sgd_logreg.predict(train_x)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.806831566548881

In [1896]:
#Predict the test values
test_y_pred = sgd_logreg.predict(test_x)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.7969821673525377

#### Therefor, When increasing the learing rate eta0=0.03 their has been and increase in the accuracy value of test and 
#### provieded us with best the accuracy value of test i.e 0.796 when compared to the previous case1

# Polynomial Regression with Degree as 3:

This is done by creating the polynomial "variables" of the existing variables, then fitting them in a regular regression model


In [1927]:
from sklearn.preprocessing import PolynomialFeatures

# Create second degree terms and interaction terms
poly_features = PolynomialFeatures(degree=3).fit(train_x)

train_x_poly = poly_features.transform(train_x)

test_x_poly = poly_features.transform(test_x)

#Mind you, this will create the polynomial terms of the categorical variables too

#if degree=3, then it creates all combinations: a, a^2, a^3, b, b^2, b^3, a.b, a^2.b, a.b^2, a^2.b^2 

In [1928]:
log_reg = LogisticRegression()

log_reg.fit(train_x_poly, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [1929]:
#Predict the train values
train_y_pred = log_reg.predict(train_x_poly)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.9411071849234394

In [1930]:
#Predict the test values
test_y_pred = log_reg.predict(test_x_poly)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.7215363511659808

## Let's regularize the polynomial regression

In [1931]:
# C is the inverse of alpha. It can only be positive
# Lower C means more regularization

log_reg = LogisticRegression(penalty='l2', 
                              C=0.01)

log_reg.fit(train_x_poly, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=0.01)

In [1932]:
#Predict the train values
train_y_pred = log_reg.predict(train_x_poly)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.8621908127208481

In [1938]:
#Predict the test values
test_y_pred = log_reg.predict(test_x_poly)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.7818930041152263

#### After changing the degree value in polynomial regression we can observe an change in the accuracy values of train and test and it is a clear indication of overfitting as there is significant difference between Train and Test accuracy values.

## LogisticRegression Model:

In [1914]:
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression(penalty='none')

log_reg.fit(train_x, train_targets)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:1173: FutureWarning: `penalty='none'`has been deprecated in 1.2 and will be removed in 1.4. To keep the past behaviour, set `penalty=None`.
  warnings.warn(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(penalty='none')

# Predicted vs. Actual values

In [1915]:
log_reg.predict(test_x)

array([1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1,
       1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1,

In [1916]:
# Create a new DataFrame

predictions = pd.DataFrame(log_reg.predict(test_x), columns=['Predicted'])

predictions

Predicted
0            1
1            1
2            0
3            1
4            1
..         ...
724          1
725          1
726          0
727          1
728          0

[729 rows x 1 columns]

In [1917]:
# Add the actual to the same DataFrame

predictions['Actual'] = np.array(test_targets)

predictions

Predicted  Actual
0            1       1
1            1       1
2            0       0
3            1       1
4            1       1
..         ...     ...
724          1       0
725          1       1
726          0       0
727          1       0
728          0       0

[729 rows x 2 columns]

# Calculate the overall accuracy

In [1918]:
from sklearn.metrics import accuracy_score

In [1919]:
#Predict the train values
train_y_pred = log_reg.predict(train_x)

#Train accuracy
accuracy_score(train_targets, train_y_pred)

0.8191990577149588

In [1920]:
#Predict the test values
test_y_pred = log_reg.predict(test_x)

#Test accuracy
accuracy_score(test_targets, test_y_pred)

0.8052126200274349

## Classification Matrix

In [1921]:
from sklearn.metrics import confusion_matrix

#We usually create the confusion matrix on test set
confusion_matrix(test_targets, test_y_pred)

array([[270,  73],
       [ 69, 317]], dtype=int64)

## Classification Report

In [1922]:
from sklearn.metrics import classification_report

#We usually create the classification report on test set
print(classification_report(test_targets, test_y_pred))

              precision    recall  f1-score   support

           0       0.80      0.79      0.79       343
           1       0.81      0.82      0.82       386

    accuracy                           0.81       729
   macro avg       0.80      0.80      0.80       729
weighted avg       0.81      0.81      0.81       729



# Discussion (3 points in total)


## List the train and test values of each model you built (1 point)

### Baseline Accuracy:
###### Train Accuracy: 0.51
###### Test Accuracy: 0.52
-----------------------------------------------------------
### SVM Model With Linear Kernel
###### Train Accuracy: 0.81
###### Test Accuracy: 0.79

### SVM Model with RBF Kernel
##### Train Accuracy:0.82
##### Test Accuracy: 0.80

### SVM Model with Poly Kernel
##### Train Accuracy: 0.83
##### Test Accuracy: 0.79

### SGD Model 1 
##### Train Accuracy: 0.79
##### Test Accuracy: 0.77

### Polynomial Regression with Degree2:
##### Train Accuracy:0.83
##### Test Accuracy: 0.78

### After Regularizing the polynomial Regression
##### Train Accuracy: 0.82
##### Test Accuracy: 0.78

----------------

### SGD Model 2 
##### Train Accuracy: 0.80
##### Test Accuracy: 0.79

### Polynomial Regression with Degree3: 
##### Train Accuracy:0.94
##### Test Accuracy: 0.72

### After Regularizing the polynomial Regression
##### Train Accuracy: 0.86
##### Test Accuracy: 0.78

------------------------------------------------------

### Logistic Regression 
##### Train Accuracy: 0.81
##### Test Accuracy: 0.80

## Which model performs the best and why? (0.5 points) How does it compare to baseline? (0.5 points)

Hint: The best model is the one that has the highest TEST score (regardless of any of the training values). If you select your model based on TRAIN values, you will lose points.

SVM Model with RBF Kernel has the best accuracy value i.e 0.80 when compared to other models. The baseline value of Test is 0.52. Baseline is used to identify the effectiveness of the model and The RBF Model test predections are above the baseline and it is considered to be a good fit. If the RBF model test accuracy was below 0.52 it will be considered as a bad model.

## Is there any evidence of overfitting in the best model, why or why not? If there is, what did you do about it? (0.5 points)

Yes, Based on the analysis we can oberver that when increasing the value of c from 0.01 to 40 and changing the gamma from auto to scale, there has been an significant difference between the test and train values of the model which can be identified as a case of overfitting.

Based on the prediction of the model, I identified when the gama was set to Auto and the C value was set to a lower value, the difference between the train and test values was reduced and provided me with the best accuracy value of test. 

## Is there any evidence of overfitting in the other models (besides the best model), why or why not? If there is, what did you do about it? (0.5 points)

 Yes, We can observe overfitting in two models which are:
 
1.  SVM Model with Poly Kernel: 

When the value of degree was increased, the difference between accuracy value of Train and Test increased when C=30,    coef=10. Therefor by analyzing the change in the parameters, and reducting the values to Degree=2, coef=0.1, c=0.1, we got the best accuracy values of Test and Train. 

2.  SGD Model 2 in Polynomial Regression when degree is set as 3:

When the value of degree was set to 3, the difference between accuracy value of Train and Test increased and it had a gap of almost 22%, but when we performed regularization on the polynomial regression the accuracy value of Test and Train reduced moreover the gap between the both value was also reduce to 8%.

